System/Posix/Files/Common.hsc

```diff
@@ -688,3 +752,398 @@ getFdPathVar (Fd fd) v =
+-- This is an extended attribute.
+-- Since Linux 5.5.
+fileVerityX                :: ExtendedFileStatus -> Bool
+-- | The  file is in the DAX (cpu direct access) state.
+-- This is an extended attribute.
+-- Since Linux 5.8.
+fileDaxX                   :: ExtendedFileStatus -> Bool
+
+-- | Checks if this file is a block device.
+isBlockDeviceX     :: ExtendedFileStatus -> Bool
+-- | Checks if this file is a character device.
+isCharacterDeviceX :: ExtendedFileStatus -> Bool
+-- | Checks if this file is a named pipe device.
+isNamedPipeX       :: ExtendedFileStatus -> Bool
+-- | Checks if this file is a regular file device.
+isRegularFileX     :: ExtendedFileStatus -> Bool
+-- | Checks if this file is a directory device.
+isDirectoryX       :: ExtendedFileStatus -> Bool
+-- | Checks if this file is a symbolic link device.
+isSymbolicLinkX    :: ExtendedFileStatus -> Bool
+-- | Checks if this file is a socket device.
+isSocketX          :: ExtendedFileStatus -> Bool
+
+isBlockDeviceX statx =
+  (fileModeX statx `intersectFileModes` fileTypeModes) == blockSpecialMode
+isCharacterDeviceX statx =
+  (fileModeX statx `intersectFileModes` fileTypeModes) == characterSpecialMode
+isNamedPipeX statx =
+  (fileModeX statx `intersectFileModes` fileTypeModes) == namedPipeMode
+isRegularFileX statx =
+  (fileModeX statx `intersectFileModes` fileTypeModes) == regularFileMode
+isDirectoryX statx =
+  (fileModeX statx `intersectFileModes` fileTypeModes) == directoryMode
+isSymbolicLinkX statx =
+  (fileModeX statx `intersectFileModes` fileTypeModes) == symbolicLinkMode
+isSocketX statx =
+  (fileModeX statx `intersectFileModes` fileTypeModes) == socketMode
+
+#if HAVE_STATX
+testFlag :: ExtendedFileStatus -> CAttributes -> Bool
+testFlag ex flag =
+  let attributes = fileAttributesX ex
+      attributes_mask = fileAttributesMaskX ex
+  -- TODO: do we need attributes_mask? Check the spec.
+  in (attributes .&. attributes_mask .&. flag) /= 0
+
+fileCompressedX ex = testFlag ex (#const STATX_ATTR_COMPRESSED)
+fileImmutableX ex = testFlag ex (#const STATX_ATTR_IMMUTABLE)
+fileAppendX ex = testFlag ex (#const STATX_ATTR_APPEND)
+fileNoDumpX ex = testFlag ex (#const STATX_ATTR_NODUMP)
+fileEncryptedX ex = testFlag ex (#const STATX_ATTR_ENCRYPTED)
+#ifdef STATX_ATTR_VERITY
+fileVerityX ex = testFlag ex (#const STATX_ATTR_VERITY)
+#else
+{-# WARNING fileVerityX "fileVerityX: not available on this platform" #-}
+fileVerityX _ = error "fileVerityX not available on this platform"
+#endif
+#ifdef STATX_ATTR_DAX
+fileDaxX ex = testFlag ex (#const STATX_ATTR_DAX)
+#else
+{-# WARNING fileDaxX "fileDaxX: not available on this platform" #-}
+fileDaxX _ = error "fileDaxX not available on this platform"
+#endif
+
+fileLinkCountX (ExtendedFileStatus statx) = unsafePerformIO $ do
+  links <- withForeignPtr statx $ (#peek struct statx, stx_nlink)
+  return $ CNlink (fromIntegral (links :: Word32))
```

Comment:
This is not an accident. We get borked results if we use `CNlink` directly:

```
$ cabal run unix-tests -- -p '/fileExtendedStatus/'                                                                               44.8s  Sat 12:45
All
  Native
    fileExtendedStatus: FAIL
      tests/FileExtendedStatus.hs:79:
      expected: 2
       but got: 4294967296002
```

Definitions are:

- statx
  - C code: `__u32`
  - Haskell type: `Word32`
- stat
  - C code: `nlink_t`
  - Haskell type: `CNlink` (wrapped `Word64`)

I'm not too sure what's going on here @vdukhovni.